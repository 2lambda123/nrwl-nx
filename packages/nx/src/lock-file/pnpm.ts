import { PackageJson } from '../utils/package-json';
import { LockFileBuilder, nodeKey } from './lock-file-builder';
import {
  LockFileGraph,
  LockFileNode,
  VersionedPackageSnapshot,
} from './utils/types';
import {
  PackageSnapshot,
  Lockfile,
  ProjectSnapshot,
} from '@pnpm/lockfile-types';
import {
  loadPnpmHoistedDepsDefinition,
  parseAndNormalizePnpmLockfile,
} from './utils/pnpm-utils';
import {
  addEdgeOuts,
  getRootVersion,
  reportUnresolvedDependencies,
  UnresolvedDependencies,
} from './utils/parsing-utils';

export function parsePnpmLockFile(lockFileContent: string): LockFileGraph {
  const builder = buildLockFileGraph(lockFileContent);
  return builder.getLockFileGraph();
}

export function prunePnpmLockFile(
  rootLockFileContent: string,
  prunedPackageJson: PackageJson
): string {
  const builder = buildLockFileGraph(rootLockFileContent);
  builder.prune(prunedPackageJson);

  return rootLockFileContent;
}

function buildLockFileGraph(content: string): LockFileBuilder {
  const data = parseAndNormalizePnpmLockfile(content);
  const hoistedDependencies = loadPnpmHoistedDepsDefinition();
  const groupedDependencies = groupDependencies(data);

  const builder = new LockFileBuilder(data.importers['.']);

  // Non-root dependencies that need to be resolved later
  // Map[packageName, specKey, PackageSnapshot]
  const unresolvedDependencies: UnresolvedDependencies<VersionedPackageSnapshot> =
    new Set<[string, string, VersionedPackageSnapshot]>();

  groupedDependencies.forEach((versionSet, packageName) => {
    versionSet.forEach(([specKey, originalKey, packageSnapshot]) => {
      const isRootVersion = isVersionHoisted(
        versionSet,
        data.importers,
        hoistedDependencies,
        packageName,
        originalKey,
        packageSnapshot
      );

      if (isRootVersion) {
        const node = parseNode(packageName, specKey, packageSnapshot, true);
        if (!builder.nodes.has(nodeKey(node))) {
          builder.addNode(node);
          builder.addEdgeIn(node, specKey);

          if (packageSnapshot.peerDependenciesMeta) {
            // for pnpm, peerDependencies are always doubled in the dependencies if installed
            addEdgeOuts({
              builder,
              node,
              section: packageSnapshot.dependencies,
              isOptionalFunc: (depName) =>
                packageSnapshot.peerDependenciesMeta[depName]?.optional,
            });
          } else {
            addEdgeOuts({
              builder,
              node,
              section: packageSnapshot.dependencies,
            });
          }
          addEdgeOuts({
            builder,
            node,
            section: packageSnapshot.optionalDependencies,
          });
        } else {
          const existingEdge = builder.nodes.get(nodeKey(node));
          builder.addEdgeIn(existingEdge, specKey);
        }
      } else {
        unresolvedDependencies.add([packageName, specKey, packageSnapshot]);
      }
    });
  });

  exhaustUnresolvedDependencies(builder, unresolvedDependencies);

  return builder;
}

function isVersionHoisted(
  versionSet: Set<[string, string, VersionedPackageSnapshot]>,
  importers: Record<string, ProjectSnapshot>,
  hoistedDependencies: Record<string, any>,
  packageName: string,
  key: string,
  packageSnapshot: PackageSnapshot
): boolean {
  // if there's only one version, it's automatically hoisted
  if (versionSet.size === 1) {
    return true;
  }
  // if dependency is defined in importers that has priority
  if (importers['.'].dependencies?.[packageName]) {
    return (
      importers['.'].dependencies[packageName] ===
      key.slice(key.lastIndexOf('/') + 1)
    );
  }
  if (importers['.'].devDependencies?.[packageName]) {
    return (
      importers['.'].devDependencies[packageName] ===
      key.slice(key.lastIndexOf('/') + 1)
    );
  }
  // some hoisted dependencies are in node_modules/{packageName} while others are in node_modules/.pnpm/{packageName}@{version}
  // modules.yaml provides better specificity than node_modules/{packageName}
  const nonVersionedKey = key.split('_')[0];
  if (
    Object.keys(hoistedDependencies).some((k) => k.startsWith(nonVersionedKey))
  ) {
    return !!hoistedDependencies[key];
  }
  const rootVersion = getRootVersion(packageName);
  return packageSnapshot.version === rootVersion;
}

function parseVersionSpec(key: string, packageName: string): string {
  if (key.startsWith(`/${packageName}/`)) {
    return key.slice(key.lastIndexOf('/') + 1);
  }
  // for alias or tarball packlage the entire key is the version spec
  return key;
}

function parseNode(
  packageName: string,
  versionSpec: string,
  snapshot: PackageSnapshot,
  isHoisted: boolean
): LockFileNode {
  const resolution = snapshot.resolution;

  let name = snapshot.name;
  // for tarball packages, the version of the snapshot is not accurate
  let version = resolution['tarball'] || snapshot.version;

  const node: LockFileNode = {
    name: name || packageName,
    ...(version && { version }),
    isHoisted,
  };

  if (!name && versionSpec.startsWith('/')) {
    node.packageName = versionSpec.slice(1, versionSpec.lastIndexOf('/'));
  }

  return node;
}

function groupDependencies(
  data: Lockfile
): Map<string, Set<[string, string, VersionedPackageSnapshot]>> {
  // packageName => Set[versionSpec, originalKey, PackageSnapshot]
  const groupedDependencies = new Map<
    string,
    Set<[string, string, VersionedPackageSnapshot]>
  >();

  Object.entries(data.packages).forEach(([key, value]) => {
    const packageName = findPackageName(key, value, data);
    const versionSpec = parseVersionSpec(key, packageName);
    const version =
      value.version ||
      versionSpec.slice(versionSpec.lastIndexOf('/') + 1).split('_')[0];
    const snapshot = { ...value, version };

    if (!groupedDependencies.has(packageName)) {
      groupedDependencies.set(
        packageName,
        new Set<[string, string, VersionedPackageSnapshot]>().add([
          versionSpec,
          key,
          snapshot,
        ])
      );
    } else {
      groupedDependencies.get(packageName).add([versionSpec, key, snapshot]);
    }
  });
  return groupedDependencies;
}

function findPackageName(
  key: string,
  value: PackageSnapshot,
  data: Lockfile
): string {
  const matchPropValue = (record: Record<string, string>): string => {
    if (!record) {
      return undefined;
    }
    const index = Object.values(record).findIndex((version) => version === key);
    if (index > -1) {
      return Object.keys(record)[index];
    }
  };

  const matchedDependencyName = (
    snapshot: Partial<PackageSnapshot>
  ): string => {
    return (
      matchPropValue(snapshot.dependencies) ||
      matchPropValue(snapshot.optionalDependencies) ||
      matchPropValue(snapshot.peerDependencies)
    );
  };

  if (value.name) {
    return value.name;
  }
  const rootDependencyName =
    matchedDependencyName(data.importers['.']) ||
    // only root importers have devDependencies
    matchPropValue(data.importers['.'].devDependencies);
  if (rootDependencyName) {
    return rootDependencyName;
  }
  const snapshots = Object.values(data.packages);
  for (let i = 0; i < snapshots.length; i++) {
    const snapshot = snapshots[i];
    const dependencyName = matchedDependencyName(snapshot);
    if (dependencyName) {
      return dependencyName;
    }
  }
  return key.startsWith('/') ? key.slice(1, key.lastIndexOf('/')) : key;
}

function exhaustUnresolvedDependencies(
  builder: LockFileBuilder,
  unresolvedDependencies: UnresolvedDependencies<PackageSnapshot>
) {
  const initialSize = unresolvedDependencies.size;
  if (!initialSize) {
    return;
  }

  unresolvedDependencies.forEach((unresolvedSet) => {
    const [packageName, versionSpec, packageSnapshot] = unresolvedSet;

    for (const n of builder.nodes.values()) {
      if (n.edgesOut && n.edgesOut.has(packageName)) {
        const edge = n.edgesOut.get(packageName);

        if (edge.versionSpec === versionSpec) {
          const node = parseNode(
            packageName,
            versionSpec,
            packageSnapshot,
            false
          );

          // we might have added the node already
          if (!builder.nodes.has(nodeKey(node))) {
            builder.addNode(node);
            builder.addEdgeIn(node, versionSpec);
            if (packageSnapshot.peerDependenciesMeta) {
              addEdgeOuts({
                builder,
                node,
                section: packageSnapshot.dependencies,
                isOptionalFunc: (depName) =>
                  packageSnapshot.peerDependenciesMeta[depName]?.optional,
              });
            } else {
              addEdgeOuts({
                builder,
                node,
                section: packageSnapshot.dependencies,
              });
            }
            addEdgeOuts({
              builder,
              node,
              section: packageSnapshot.optionalDependencies,
            });
          } else {
            const existingNode = builder.nodes.get(nodeKey(node));
            builder.addEdgeIn(existingNode, versionSpec);
          }
          unresolvedDependencies.delete(unresolvedSet);
          return;
        }
      }
    }
  });

  if (initialSize === unresolvedDependencies.size) {
    // ignore packages that were parsed already with different hashmap
    unresolvedDependencies.forEach((unresolvedDependency) => {
      const [packageName, versionSpec] = unresolvedDependency;
      if (builder.nodes.has(`${packageName}@${versionSpec.split('_')[0]}`)) {
        unresolvedDependencies.delete(unresolvedDependency);
      }
    });
  }

  if (initialSize === unresolvedDependencies.size) {
    reportUnresolvedDependencies(unresolvedDependencies);
  } else if (unresolvedDependencies.size > 0) {
    exhaustUnresolvedDependencies(builder, unresolvedDependencies);
  }
}
