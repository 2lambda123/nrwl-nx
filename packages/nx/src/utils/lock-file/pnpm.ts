import {
  LockFileData,
  PackageDependency,
  PackageVersions,
} from './lock-file-type';
import { load, dump } from '@zkochan/js-yaml';
import {
  sortObject,
  hashString,
  isRootVersion,
  TransitiveLookupFunctionInput,
  generatePrunnedHash,
} from './utils';
import { satisfies } from 'semver';
import { dematerialize } from 'rxjs/operators';

type PackageMeta = {
  key: string;
  specifier?: string;
  isDevDependency?: boolean;
  isDependency?: boolean;
  dependencyDetails: Record<string, Record<string, string>>;
  dev?: boolean;
  optional?: boolean;
  peer?: boolean;
};

type Dependencies = Record<string, Omit<PackageDependency, 'packageMeta'>>;

type VersionInfoWithInlineSpecifier = {
  version: string;
  specifier: string;
};

type PnpmLockFile = {
  lockfileVersion: number;
  specifiers?: Record<string, string>;
  dependencies?: Record<
    string,
    string | { version: string; specifier: string }
  >;
  devDependencies?: Record<
    string,
    string | { version: string; specifier: string }
  >;
  packages?: Dependencies;
  importers?: Record<string, { specifiers: Record<string, string> }>;

  time?: Record<string, string>; // e.g.   /@babel/core/7.19.6: '2022-10-20T09:03:36.074Z'
  overrides?: Record<string, string>; // js-yaml@^4.0.0: npm:@zkochan/js-yaml@0.0.6
  patchedDependencies?: Record<string, { path: string; hash: string }>; // e.g.  pkg@5.7.0: { path: 'patches/pkg@5.7.0.patch', hash: 'sha512-...' }
  neverBuiltDependencies?: string[]; // e.g.  ['core-js', 'level']
  onlyBuiltDependencies?: string[]; // e.g.  ['vite']
  packageExtensionsChecksum?: string; // e.g.  'sha512-...' DO NOT COPY TO PRUNED LOCKFILE
};

const LOCKFILE_YAML_FORMAT = {
  blankLines: true,
  lineWidth: 1000,
  noCompatMode: true,
  noRefs: true,
  sortKeys: false,
};

/**
 * Parses pnpm-lock.yaml file to `LockFileData` object
 *
 * @param lockFile
 * @returns
 */
export function parsePnpmLockFile(lockFile: string): LockFileData {
  const data: PnpmLockFile = load(lockFile);
  const { dependencies, devDependencies, packages, specifiers, ...metadata } =
    data;

  return {
    dependencies: mapPackages(
      dependencies,
      devDependencies,
      specifiers,
      packages,
      metadata.lockfileVersion.toString().endsWith('inlineSpecifiers')
    ),
    lockFileMetadata: { ...metadata },
    hash: hashString(lockFile),
  };
}

function mapPackages(
  dependencies: Record<string, string | VersionInfoWithInlineSpecifier>,
  devDependencies: Record<string, string | VersionInfoWithInlineSpecifier>,
  specifiers: Record<string, string>,
  packages: Dependencies,
  inlineSpecifiers: boolean
): LockFileData['dependencies'] {
  const mappedPackages: LockFileData['dependencies'] = {};

  Object.entries(packages).forEach(([key, value]) => {
    // construct packageMeta object
    const meta = mapMetaInformation(
      { dependencies, devDependencies, specifiers },
      inlineSpecifiers,
      key,
      value
    );

    // create new key
    const version = key.split('/').pop().split('_')[0];
    const packageName = key.slice(1, key.lastIndexOf('/'));
    const newKey = `${packageName}@${version}`;

    if (!mappedPackages[packageName]) {
      mappedPackages[packageName] = {};
    }
    if (mappedPackages[packageName][newKey]) {
      mappedPackages[packageName][newKey].packageMeta.push(meta);
    } else {
      const { dev, optional, peer, ...rest } = value;
      mappedPackages[packageName][newKey] = {
        ...rest,
        version,
        packageMeta: [meta],
      };
    }
  });
  Object.keys(mappedPackages).forEach((packageName) => {
    const versions = mappedPackages[packageName];
    const versionKeys = Object.keys(versions);

    if (versionKeys.length === 1) {
      versions[versionKeys[0]].rootVersion = true;
    } else {
      const rootVersionKey = versionKeys.find((v) =>
        isRootVersion(packageName, versions[v].version)
      );
      // this should never happen, but just in case
      if (rootVersionKey) {
        versions[rootVersionKey].rootVersion = true;
      }
    }
  });

  return mappedPackages;
}

// maps packageMeta based on dependencies, devDependencies and (inline) specifiers
function mapMetaInformation(
  {
    dependencies,
    devDependencies,
    specifiers,
  }: Omit<PnpmLockFile, 'lockfileVersion' | 'packages'>,
  hasInlineSpefiers,
  key: string,
  {
    dev,
    optional,
    peer,
    ...dependencyDetails
  }: Omit<PackageDependency, 'packageMeta'>
): PackageMeta {
  const matchingVersion = key.split('/').pop();
  const packageName = key.slice(1, key.lastIndexOf('/'));

  const isDependency = isVersionMatch(
    dependencies?.[packageName],
    matchingVersion,
    hasInlineSpefiers
  );
  const isDevDependency = isVersionMatch(
    devDependencies?.[packageName],
    matchingVersion,
    hasInlineSpefiers
  );

  let specifier;
  if (isDependency || isDevDependency) {
    if (hasInlineSpefiers) {
      specifier =
        getSpecifier(dependencies?.[packageName]) ||
        getSpecifier(devDependencies?.[packageName]);
    } else {
      if (isDependency || isDevDependency) {
        specifier = specifiers[packageName];
      }
    }
  }

  return {
    key,
    isDependency,
    isDevDependency,
    specifier,
    ...(dev && { dev }),
    ...(optional && { optional }),
    ...(peer && { peer }),
    dependencyDetails: {
      ...(dependencyDetails.dependencies && {
        dependencies: dependencyDetails.dependencies,
      }),
      ...(dependencyDetails.peerDependencies && {
        peerDependencies: dependencyDetails.peerDependencies,
      }),
      ...(dependencyDetails.optionalDependencies && {
        optionalDependencies: dependencyDetails.optionalDependencies,
      }),
      ...(dependencyDetails.transitivePeerDependencies && {
        transitivePeerDependencies:
          dependencyDetails.transitivePeerDependencies,
      }),
    },
  };
}

// version match for dependencies w/ or w/o inline specifier
function isVersionMatch(
  versionInfo: string | { version: string; specifier: string },
  matchingVersion,
  hasInlineSpefiers
): boolean {
  if (!versionInfo) {
    return false;
  }
  if (!hasInlineSpefiers) {
    return versionInfo === matchingVersion;
  }

  return (
    (versionInfo as VersionInfoWithInlineSpecifier).version === matchingVersion
  );
}

function getSpecifier(
  versionInfo: string | { version: string; specifier: string }
): string {
  return (
    versionInfo && (versionInfo as VersionInfoWithInlineSpecifier).specifier
  );
}

/**
 * Generates pnpm-lock.yml file from `LockFileData` object
 *
 * @param lockFile
 * @returns
 */
export function stringifyPnpmLockFile(lockFileData: LockFileData): string {
  const pnpmLockFile = unmapLockFile(lockFileData);

  return dump(pnpmLockFile, LOCKFILE_YAML_FORMAT);
}

// revert lock file to it's original state
function unmapLockFile(lockFileData: LockFileData): PnpmLockFile {
  const devDependencies: Record<
    string,
    string | VersionInfoWithInlineSpecifier
  > = {};
  const dependencies: Record<string, string | VersionInfoWithInlineSpecifier> =
    {};
  const packages: Dependencies = {};
  const specifiers: Record<string, string> = {};
  const inlineSpecifiers = lockFileData.lockFileMetadata.lockfileVersion
    .toString()
    .endsWith('inlineSpecifiers');

  const packageNames = Object.keys(lockFileData.dependencies);
  for (let i = 0; i < packageNames.length; i++) {
    const packageName = packageNames[i];
    const versions = Object.values(lockFileData.dependencies[packageName]);

    versions.forEach(({ packageMeta, version: _, rootVersion, ...rest }) => {
      (packageMeta as PackageMeta[]).forEach(
        ({
          key,
          specifier,
          isDependency,
          isDevDependency,
          dependencyDetails,
          dev,
          optional,
          peer,
        }) => {
          let version;
          if (inlineSpecifiers) {
            version = {
              specifier,
              version: key.slice(key.lastIndexOf('/') + 1),
            };
          } else {
            version = key.slice(key.lastIndexOf('/') + 1);

            if (specifier) {
              specifiers[packageName] = specifier;
            }
          }

          if (isDependency) {
            dependencies[packageName] = version;
          }
          if (isDevDependency) {
            devDependencies[packageName] = version;
          }
          packages[key] = {
            ...rest,
            dev: !!dev,
            ...(optional && { optional }),
            ...(peer && { peer }),
            ...dependencyDetails,
          };
        }
      );
    });
  }

  const { time, ...lockFileMetatada } = lockFileData.lockFileMetadata as Omit<
    PnpmLockFile,
    'specifiers' | 'importers' | 'devDependencies' | 'dependencies' | 'packages'
  >;

  return {
    ...(lockFileMetatada as { lockfileVersion: number }),
    specifiers: sortObject(specifiers),
    dependencies: sortObject(dependencies),
    devDependencies: sortObject(devDependencies),
    packages: sortObject(packages),
    time,
  };
}

/**
 * Returns matching version of the dependency
 */
export function transitiveDependencyPnpmLookup({
  packageName,
  versions,
  version,
}: TransitiveLookupFunctionInput): PackageDependency {
  const combinedDependency = Object.values(versions).find((v) =>
    v.packageMeta.some((m) => m.key === `/${packageName}/${version}`)
  );
  if (combinedDependency) {
    return combinedDependency;
  }
  // pnpm's dependencies always point to the exact version so this block is only for insurrance
  return Object.values(versions).find((v) => satisfies(v.version, version));
}

/**
 * Prunes the lock file data based on the list of packages and their transitive dependencies
 *
 * @param lockFileData
 * @returns
 */
export function prunePnpmLockFile(
  lockFileData: LockFileData,
  packages: string[],
  projectName?: string
): LockFileData {
  const dependencies = pruneDependencies(
    lockFileData.dependencies,
    packages,
    projectName
  );
  const prunedLockFileData = {
    lockFileMetadata: pruneMetadata(
      lockFileData.lockFileMetadata,
      dependencies
    ),
    dependencies,
    hash: generatePrunnedHash(lockFileData.hash, packages, projectName),
  };
  return prunedLockFileData;
}

// iterate over packages to collect the affected tree of dependencies
function pruneDependencies(
  dependencies: LockFileData['dependencies'],
  packages: string[],
  projectName?: string
): LockFileData['dependencies'] {
  const result: LockFileData['dependencies'] = {};

  packages.forEach((packageName) => {
    if (dependencies[packageName]) {
      const [key, { packageMeta, ...value }] = Object.entries(
        dependencies[packageName]
      ).find(([_, v]) => v.rootVersion);
      result[packageName] = result[packageName] || {};
      const metaKey = `/${packageName}/${value.version}`;
      const meta = packageMeta.find((m) => m.key.startsWith(metaKey));

      result[packageName][key] = Object.assign(value, {
        packageMeta: [
          {
            isDependency: true,
            key: meta.key,
            specifier: value.version,
            dependencyDetails: meta.dependencyDetails,
          },
        ],
      });

      pruneTransitiveDependencies(
        [packageName],
        dependencies,
        result,
        result[packageName][key]
      );
    } else {
      console.warn(
        `Could not find ${packageName} in the lock file. Skipping...`
      );
    }
  });

  return result;
}

function pruneMetadata(
  lockFileMetadata: LockFileData['lockFileMetadata'],
  prunedDependencies: Record<string, PackageVersions>
): LockFileData['lockFileMetadata'] {
  // These should be removed from the lock file metadata since we don't have them in the package.json
  // overrides, patchedDependencies, neverBuiltDependencies, onlyBuiltDependencies, packageExtensionsChecksum
  return {
    lockfileVersion: lockFileMetadata.lockfileVersion,
    ...(lockFileMetadata.time && {
      time: pruneTime(lockFileMetadata.time, prunedDependencies),
    }),
  };
}

function pruneTime(
  time: Record<string, string>,
  prunedDependencies: Record<string, PackageVersions>
): Record<string, string> {
  const result: Record<string, string> = {};

  Object.entries(time).forEach(([key, value]) => {
    const packageName = key.slice(1, key.lastIndexOf('/'));
    const version = key.slice(key.lastIndexOf('/'));
    if (
      prunedDependencies[packageName] &&
      prunedDependencies[packageName][`${packageName}@${version}`]
    ) {
      result[key] = value;
    }
  });

  return result;
}

// find all transitive dependencies of already pruned packages
// and adds them to the collection
// recursively prune their dependencies
function pruneTransitiveDependencies(
  parentPackages: string[],
  dependencies: LockFileData['dependencies'],
  prunedDeps: LockFileData['dependencies'],
  parent: PackageDependency
): void {
  Object.entries({
    ...parent.dependencies,
    ...parent.optionalDependencies,
  }).forEach(([packageName, version]: [string, string]) => {
    const versions = dependencies[packageName];
    if (versions) {
      const dependency = transitiveDependencyPnpmLookup({
        packageName,
        parentPackages,
        versions,
        version,
      });
      if (dependency) {
        if (!prunedDeps[packageName]) {
          prunedDeps[packageName] = {};
        }
        const { packageMeta, rootVersion, ...value } = dependency;
        const key = `${packageName}@${value.version}`;
        const meta = findPackageMeta(packageMeta, parentPackages);
        if (prunedDeps[packageName][key]) {
          // TODO not sure if this is important?
        } else {
          const packageMeta: PackageMeta = {
            key: meta.key,
            dependencyDetails: meta.dependencyDetails,
          };
          prunedDeps[packageName][key] = Object.assign(value, {
            rootVersion: false,
            packageMeta: [packageMeta],
          });
          if (parent.optionalDependencies?.[packageName]) {
            packageMeta.optional = true;
          }
          pruneTransitiveDependencies(
            [...parentPackages, packageName],
            dependencies,
            prunedDeps,
            prunedDeps[packageName][key]
          );
        }
      }
    }
  });
}

function findPackageMeta(
  packageMeta: PackageMeta[],
  parentPackages: string[]
): PackageMeta {
  const matchPackageVersionModifier =
    (version: string) => (packageName: string) => {
      const normalizedName = packageName.split('/').join('+');
      if (version.includes(`_${normalizedName}@`)) {
        return true;
      }
    };

  const nestedDependency =
    packageMeta.find((m) => {
      const version = m.key.split('/').pop();
      // it's modified by a single dependency
      // e.g. /@org/my-package/1.0.0_@babel+core@1.2.3
      return (
        version.includes('_') &&
        parentPackages.some(matchPackageVersionModifier(version))
      );
    }) ||
    // it's not modified by a single dependency but a mix of them
    // e.g. /@org/my-package/1.0.0_asfgasgasgasg
    packageMeta.find((m) => m.key.split('/').pop().includes('_'));
  if (nestedDependency) {
    return nestedDependency;
  }
  // otherwise it's a root dependency
  return packageMeta.find((m) => !m.key.split('/').pop().includes('_'));
}
