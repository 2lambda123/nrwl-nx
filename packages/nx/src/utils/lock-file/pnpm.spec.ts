import { parseLockFile, stringifyLockFile } from './pnpm';
import {
  lockFile,
  lockFileWithInlineSpecifiers,
} from './__fixtures__/pnpm.lock';

describe('pnpm LockFile utility', () => {
  describe('standard lock file', () => {
    const parsedLockFile = parseLockFile(lockFile);

    it('should parse lockfile correctly', () => {
      expect(parsedLockFile.lockFileMetadata).toEqual({ lockfileVersion: 5.4 });
      expect(Object.keys(parsedLockFile.dependencies).length).toEqual(349);
      expect(
        parsedLockFile.dependencies['@ampproject/remapping@2.2.0']
      ).toMatchSnapshot();
      expect(parsedLockFile.dependencies['typescript@4.8.3']).toMatchSnapshot();
    });

    it('should map various versions of packages', () => {
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping@0.1.1']
      ).toBeDefined();
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping@0.3.2']
      ).toBeDefined();
    });

    it('should map various instances of the same version', () => {
      expect(
        parsedLockFile.dependencies['jest-pnp-resolver@1.2.2'].packageMeta
          .length
      ).toEqual(2);
      expect(
        parsedLockFile.dependencies['jest-pnp-resolver@1.2.2'].packageMeta[0]
          .key
      ).toEqual('/jest-pnp-resolver/1.2.2_jest-resolve@28.1.1');
      expect(
        parsedLockFile.dependencies['jest-pnp-resolver@1.2.2'].packageMeta[1]
          .key
      ).toEqual('/jest-pnp-resolver/1.2.2_jest-resolve@28.1.3');

      expect(
        parsedLockFile.dependencies['jest-pnp-resolver@1.2.2'].packageMeta[0]
          .dependencyDetails.dependencies
      ).toEqual({ 'jest-resolve': '28.1.1' });
      expect(
        parsedLockFile.dependencies['jest-pnp-resolver@1.2.2'].packageMeta[1]
          .dependencyDetails.dependencies
      ).toEqual({ 'jest-resolve': '28.1.3' });
    });

    it('should properly extract specifier', () => {
      expect(
        parsedLockFile.dependencies['@ampproject/remapping@2.2.0']
          .packageMeta[0].specifier
      ).toBeUndefined();
      expect(
        parsedLockFile.dependencies['typescript@4.8.3'].packageMeta[0].specifier
      ).toEqual('~4.8.2');
    });

    it('should properly extract dev dependency', () => {
      expect(
        parsedLockFile.dependencies['@ampproject/remapping@2.2.0']
          .packageMeta[0].isDevDependency
      ).toEqual(false);
      expect(
        parsedLockFile.dependencies['typescript@4.8.3'].packageMeta[0]
          .isDevDependency
      ).toEqual(true);
    });

    it('should match the original file on stringification', () => {
      expect(stringifyLockFile(parsedLockFile)).toEqual(lockFile);
    });
  });

  describe('lock file with inline specifiers', () => {
    const parsedLockFile = parseLockFile(lockFileWithInlineSpecifiers);

    it('should parse lockfile (IS)', () => {
      expect(parsedLockFile.lockFileMetadata).toEqual({
        lockfileVersion: '5.4-inlineSpecifiers',
      });
      expect(Object.keys(parsedLockFile.dependencies).length).toEqual(349);
      expect(
        parsedLockFile.dependencies['@ampproject/remapping@2.2.0']
      ).toMatchSnapshot();
      expect(parsedLockFile.dependencies['typescript@4.8.3']).toMatchSnapshot();
    });

    it('should map various versions of packages (IS)', () => {
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping@0.1.1']
      ).toBeDefined();
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping@0.3.2']
      ).toBeDefined();
    });

    it('should map various instances of the same version (IS)', () => {
      expect(
        parsedLockFile.dependencies['jest-pnp-resolver@1.2.2'].packageMeta
          .length
      ).toEqual(2);
      expect(
        parsedLockFile.dependencies['jest-pnp-resolver@1.2.2'].packageMeta[0]
          .key
      ).toEqual('/jest-pnp-resolver/1.2.2_jest-resolve@28.1.1');
      expect(
        parsedLockFile.dependencies['jest-pnp-resolver@1.2.2'].packageMeta[1]
          .key
      ).toEqual('/jest-pnp-resolver/1.2.2_jest-resolve@28.1.3');

      expect(
        parsedLockFile.dependencies['jest-pnp-resolver@1.2.2'].packageMeta[0]
          .dependencyDetails.dependencies
      ).toEqual({ 'jest-resolve': '28.1.1' });
      expect(
        parsedLockFile.dependencies['jest-pnp-resolver@1.2.2'].packageMeta[1]
          .dependencyDetails.dependencies
      ).toEqual({ 'jest-resolve': '28.1.3' });
    });

    it('should properly extract specifier (IS)', () => {
      expect(
        parsedLockFile.dependencies['@ampproject/remapping@2.2.0']
          .packageMeta[0].specifier
      ).toBeUndefined();
      expect(
        parsedLockFile.dependencies['typescript@4.8.3'].packageMeta[0].specifier
      ).toEqual('~4.8.2');
    });

    it('should properly extract dev dependency (IS)', () => {
      expect(
        parsedLockFile.dependencies['@ampproject/remapping@2.2.0']
          .packageMeta[0].isDevDependency
      ).toEqual(false);
      expect(
        parsedLockFile.dependencies['typescript@4.8.3'].packageMeta[0]
          .isDevDependency
      ).toEqual(true);
    });

    it('should match the original file on stringification (IS)', () => {
      expect(stringifyLockFile(parsedLockFile)).toEqual(
        lockFileWithInlineSpecifiers
      );
    });
  });
});
