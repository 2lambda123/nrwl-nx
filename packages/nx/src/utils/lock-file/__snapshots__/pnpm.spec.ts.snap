// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pnpm LockFile utility lock file with inline specifiers should parse lockfile (IS) 1`] = `
Object {
  "@ampproject/remapping@2.2.0": Object {
    "dependencies": Object {
      "@jridgewell/gen-mapping": "0.1.1",
      "@jridgewell/trace-mapping": "0.3.15",
    },
    "dev": true,
    "engines": Object {
      "node": ">=6.0.0",
    },
    "packageMeta": Array [
      Object {
        "dependencyDetails": Object {
          "dependencies": Object {
            "@jridgewell/gen-mapping": "0.1.1",
            "@jridgewell/trace-mapping": "0.3.15",
          },
        },
        "isDependency": false,
        "isDevDependency": false,
        "key": "/@ampproject/remapping/2.2.0",
        "specifier": undefined,
      },
    ],
    "resolution": Object {
      "integrity": "sha512-qRmjj8nj9qmLTQXXmaR1cck3UXSRMPrbsLJAasZpF+t3riI71BXed5ebIOYwQntykeZuhjsdweEc9BxH5Jc26w==",
    },
    "rootVersion": true,
    "version": "2.2.0",
  },
}
`;

exports[`pnpm LockFile utility lock file with inline specifiers should parse lockfile (IS) 2`] = `
Object {
  "typescript@4.8.3": Object {
    "dev": true,
    "engines": Object {
      "node": ">=4.2.0",
    },
    "hasBin": true,
    "packageMeta": Array [
      Object {
        "dependencyDetails": Object {},
        "isDependency": false,
        "isDevDependency": true,
        "key": "/typescript/4.8.3",
        "specifier": "~4.8.2",
      },
    ],
    "resolution": Object {
      "integrity": "sha512-goMHfm00nWPa8UvR/CPSvykqf6dVV8x/dp0c5mFTMTIu0u0FlGWRioyy7Nn0PGAdHxpJZnuO/ut+PpQ8UiHAig==",
    },
    "rootVersion": true,
    "version": "4.8.3",
  },
}
`;

exports[`pnpm LockFile utility standard lock file should parse lockfile correctly 1`] = `
Object {
  "@ampproject/remapping@2.2.0": Object {
    "dependencies": Object {
      "@jridgewell/gen-mapping": "0.1.1",
      "@jridgewell/trace-mapping": "0.3.15",
    },
    "dev": true,
    "engines": Object {
      "node": ">=6.0.0",
    },
    "packageMeta": Array [
      Object {
        "dependencyDetails": Object {
          "dependencies": Object {
            "@jridgewell/gen-mapping": "0.1.1",
            "@jridgewell/trace-mapping": "0.3.15",
          },
        },
        "isDependency": false,
        "isDevDependency": false,
        "key": "/@ampproject/remapping/2.2.0",
        "specifier": undefined,
      },
    ],
    "resolution": Object {
      "integrity": "sha512-qRmjj8nj9qmLTQXXmaR1cck3UXSRMPrbsLJAasZpF+t3riI71BXed5ebIOYwQntykeZuhjsdweEc9BxH5Jc26w==",
    },
    "rootVersion": true,
    "version": "2.2.0",
  },
}
`;

exports[`pnpm LockFile utility standard lock file should parse lockfile correctly 2`] = `
Object {
  "typescript@4.8.3": Object {
    "dev": true,
    "engines": Object {
      "node": ">=4.2.0",
    },
    "hasBin": true,
    "packageMeta": Array [
      Object {
        "dependencyDetails": Object {},
        "isDependency": false,
        "isDevDependency": true,
        "key": "/typescript/4.8.3",
        "specifier": "~4.8.2",
      },
    ],
    "resolution": Object {
      "integrity": "sha512-goMHfm00nWPa8UvR/CPSvykqf6dVV8x/dp0c5mFTMTIu0u0FlGWRioyy7Nn0PGAdHxpJZnuO/ut+PpQ8UiHAig==",
    },
    "rootVersion": true,
    "version": "4.8.3",
  },
}
`;
